File: DartPointTracker/Models/Game.cs
namespace DartPointTracker.Models;
public class Game(List<Player> players, int gamePoints) {
    private string GameId { get; set; } = Guid.NewGuid().ToString();
    public List<Player> Players { get; set; } = players.Select(player => {
        player.Score = gamePoints;
        return player;
    }).ToList();
    public List<Player> Ranking { get; set; } = [];
    public Player CurrentPlayer { get; set; } = players[0];
    public bool IsFinished { get; set; }
    public bool GameSaved { get; set; }
    private int TotalThrows { get; set; } = 0;
    public void AddDartThrow(int points) {
        TotalThrows++;
        var (validThrow, currentRoundThrowNumber) = CurrentPlayer.AddDartThrow(points);
        if (CurrentPlayer.Won) {
            Ranking.Add(CurrentPlayer);
            CurrentPlayer.ThrowNumberInGameAtWin = TotalThrows;
            if (Players.Count(player => player.Won) == Players.Count - 1) {
                IsFinished = true;
                Ranking.Add(Players.First(player => !player.Won));
                return; }
            NextPlayer(); }
        else if (!validThrow || currentRoundThrowNumber == 0) {
            NextPlayer(); } }
    public void RemoveLastDartThrow() {
        if (TotalThrows > 0) {
            TotalThrows--; }
        else {
            return; }
        if (CurrentPlayer.CurrentRoundThrowNumber == 0) {
            CurrentPlayer.ReloadPreviousRound();
            PreviousPlayer(); }
        while (CurrentPlayer.Won) {
            if (CurrentPlayer.ThrowNumberInGameAtWin == TotalThrows + 1) {
                CurrentPlayer.Won = false;
                break; }
            PreviousPlayer(); }
        CurrentPlayer.RemoveLastDartThrow(); }
    private void PreviousPlayer() {
        var index = Players.IndexOf(CurrentPlayer);
        CurrentPlayer = index == 0 ? Players[^1] : Players[index - 1]; }
    private void NextPlayer() {
        do {
            var index = Players.IndexOf(CurrentPlayer);
            CurrentPlayer = index == Players.Count - 1 ? Players[0] : Players[index + 1];
        } while (CurrentPlayer.Won);
        CurrentPlayer.StartNewRound(); } }

File: dart-point-tracker-vue-ts/src/models/game.ts
import Player from './player'; 

class Game {
  gameId: string;
  gamePoints: number;
  players: Player[];
  ranking: Player[];
  currentPlayer: Player;
  isFinished: boolean;
  gameSaved: boolean;
  private totalThrows: number;

  constructor(players: Player[], gamePoints: number) {
    this.gameId = (<any>crypto).randomUUID(); 
    this.gamePoints = gamePoints;
    this.players = players.map(player => {
      player.score = gamePoints;
      return player;
    });
    this.ranking = [];
    this.currentPlayer = players[0];
    this.isFinished = false;
    this.gameSaved = false;
    this.totalThrows = 0;
  }

  addDartThrow(points: number): void {
    this.totalThrows++;
    const { validThrow, currentRoundThrowNumber } = this.currentPlayer.addDartThrow(points);

    if (this.currentPlayer.won) {
      this.ranking.push(this.currentPlayer);
      this.currentPlayer.throwNumberInGameAtWin = this.totalThrows;
      if (this.players.filter(player => player.won).length === this.players.length - 1) {
        this.isFinished = true;
        this.ranking.push(this.players.find(player => !player.won)!);
        return;
      }
      this.nextPlayer();
    } else if (!validThrow || currentRoundThrowNumber === 0) {
      this.nextPlayer();
    }
  }

  removeLastDartThrow(): void {
    if (this.totalThrows > 0) {
      this.totalThrows--;
    } else {
      return;
    }

    if (this.currentPlayer.currentRoundThrowNumber === 0) {
      this.currentPlayer.reloadPreviousRound();
      this.previousPlayer();
    }

    while (this.currentPlayer.won) {
      if (this.currentPlayer.throwNumberInGameAtWin === this.totalThrows + 1) {
        this.currentPlayer.won = false;
        break;
      }
      this.previousPlayer();
    }

    this.currentPlayer.removeLastDartThrow();
  }

  private previousPlayer(): void {
    const index = this.players.indexOf(this.currentPlayer);
    this.currentPlayer = index === 0 ? this.players[this.players.length - 1] : this.players[index - 1];
  }

  private nextPlayer(): void {
    do {
      const index = this.players.indexOf(this.currentPlayer);
      this.currentPlayer = index === this.players.length - 1 ? this.players[0] : this.players[index + 1];
    } while (this.currentPlayer.won);

    this.currentPlayer.startNewRound();
  }
}

export default Game;
